File: Python3.x/21-Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = ListNode(None)
        point = head
        while l1 and l2:
            if l1.val > l2.val:  # 处理小的l2
                tmp = ListNode(l2.val)
                point.next = tmp
                point, l2 = point.next, l2.next
            else:  # 处理小的或者相等的l1
                tmp = ListNode(l1.val)
                point.next = tmp
                point, l1 = point.next, l1.next
        tmp_point = l1 if l1 else l2  # 将剩余的链表直接添加到合并链表的尾部
        point.next = tmp_point
        return head.next

File: Algorithms/C++/21-Merge_Two_Sorted_Lists.cpp
// Runtime: 8 ms, faster than 82.61% of C++ online submissions for Merge Two Sorted Lists.
// Memory Usage: 9 MB, less than 28.99% of C++ online submissions for Merge Two Sorted Lists.
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode dummy(0);
        ListNode* tail = &dummy;
        while (l1 && l2) {
            ListNode*& node = l1->val < l2->val ? l1 : l2;
            tail = tail->next = node;
            node = node->next;
        }
        tail->next = l1 ? l1 : l2;
        return dummy.next;
    }
};